### PR TITLE
build: Upgrade pipelines to Node16 + upgrade `nvm` recommendation

### PR DESCRIPTION
**Pull request by andre4i**
## Description

Refreshing https://github.com/microsoft/FluidFramework/pull/13807

Update the recommended versions of Node.js for developers to use, as well as the version used by CI.

This updates from version 14 (which goes out of support on 2023-04-30) to version 16 (which goes out of support on 2023-09-11).

The minimum, as enforced by the root `package.json`'s `"engines"` block is not updated to 16 yet. This can be done after a migration period.

This has no impact on the produced packages, but will allow (but not yet require) developers working in this repo to upgrade.

Users of [nvm](https://npm.github.io/installation-setup-docs/installing/using-a-node-version-manager.html) can upgrade by running: `nvm use && rm -r ./node_modules/ && pnpm i` in the root of the repo. Deleating the node_modules directory is needed to trigger reinstalling of nodegit to avoid an "Error: Module did not self-register" when building our codebase.

Some of the packages still use `npm` (and not `pnpm`). If working on those packages, you may want to downgrade `npm` to version 6 after upgrading node so that you avoid converting package-lock files to the new format unexpectedly.

## Reviewer Guidance

Feedback on this process, as well as the actual implementations / diff is encouraged.

Questions:
1. Do we want to wait for the rest of the usages of npm to migrate to pnpm? (Perhaps before this, or maybe just before updating the minimum version?)
2. Is there anywhere in our docs that needs updating?

### CODE REVIEW

**Comment by CraigMacomber** [comment]
I think https://github.com/microsoft/FluidFramework/pull/13807 Has a better description (it explains why you have to remove node_modules, links to nvm docs, notes the relevant date etc.). I recommend including more of the details from it. Also I think you missed a  "`" in the last line.

File: pnpm-lock.yaml [hunk 1]
```diff
@@ -31,9 +33,11 @@ importers:
       '@fluidframework/test-tools': 0.2.3074
       '@microsoft/api-documenter': 7.17.9
       '@microsoft/api-extractor': 7.22.2
+      '@octokit/core': 4.2.0
       c8: 7.7.1
       concurrently: 6.3.0
       copyfiles: 2.4.1
+      danger: 10.9.0_@octokit+core@4.2.0
```
Comment (CraigMacomber) [inline pnpm-lock.yaml:40]: Since we can't include a comment about the issue here in this file, at the very least you should note it in the PR description that this is a work around and not a proper fix, and probably include a TODO comment in `build-tools/packages/build-cli/src/commands/run/bundleStats.ts` noting the issue as well.

File: package.json [hunk 2]
```diff
@@ -129,9 +129,11 @@
 		"@fluidframework/test-tools": "^0.2.3074",
 		"@microsoft/api-documenter": "^7.17.9",
 		"@microsoft/api-extractor": "^7.22.2",
+		"@octokit/core": "^4.0.5",
 		"c8": "^7.7.1",
 		"concurrently": "^6.2.0",
 		"copyfiles": "^2.4.1",
+		"danger": "^10.9.0",
```
Comment (CraigMacomber) [inline package.json:136]: Since we can't include a comment in here, you should note that this is a workaround in the pPR description and include a TODO comment about it in build-tools/packages/build-cli/src/commands/run/bundleStats.ts

File: package.json [hunk 3]
```diff
@@ -129,9 +129,11 @@
 		"@fluidframework/test-tools": "^0.2.3074",
 		"@microsoft/api-documenter": "^7.17.9",
 		"@microsoft/api-extractor": "^7.22.2",
+		"@octokit/core": "^4.0.5",
```
Comment (CraigMacomber) [inline package.json:132]: It would be good to explain why this change is needed somewhere.

Comment (andre4i) [reply package.json:132]: We can't add comments in here. This is needed as per pnpm's warning of this being a missing peer dependency.

**Comment by andre4i** [comment]
Hey @tylerbutler 

Any ideas why this is failing?

```
Run JulienKode/pull-request-name-linter-action@8c05fb989d9f15[6](https://github.com/microsoft/FluidFramework/actions/runs/4410527518/jobs/7728089296#step:5:7)ce61e33754f9802c9d3cffa58
Error: Error: subject-empty:subject may not be empty
type-empty:type may not be empty
Error: subject-empty:subject may not be empty
type-empty:type may not be empty
```

**Review by tylerbutler — COMMENTED** [review]
Changes look good. I would like the server folks to take a look too because this change may impact some of their workflows. I suggest posting to the Teams channel and I can mention specific folks if needed.

**Comment by ChumpChief** [comment]
> Some of the packages still use npm (and not pnpm). If working on those packages, you may want to downgrade npm to version 6 after upgrading node so that you avoid converting package-lock files to the new format unexpectedly.

It seems like it would make more sense to just upgrade the lockfiles - is there any reason to keep them in old format after this upgrade?  I can imagine that pnpm would make this moot but in the meantime seems fine to move forward?

**Comment by CraigMacomber** [comment]
> > Some of the packages still use npm (and not pnpm). If working on those packages, you may want to downgrade npm to version 6 after upgrading node so that you avoid converting package-lock files to the new format unexpectedly.
> 
> It seems like it would make more sense to just upgrade the lockfiles - is there any reason to keep them in old format after this upgrade? I can imagine that pnpm would make this moot but in the meantime seems fine to move forward?

We are not upgrading the minimum supported version of the tools yet. People modifying dependencies of packages using npm can simply not upgrade node yet or can upgrade then downgrade npm (or can upgrade their package when they modify it). Since using old npm with new node is more supported than using new npm with old node (at least I think this is the case) this seems like a better incremental approach than the other way (update npm first).

Instead of being incremental we could force everyone to upgrade node at the exact same time, and also upgrade npm and the lock file. This seems risky as it would be a large source change (a pain to revert) and likely to break someone (and thus need to be reverted). This seems like a worse option.

Thus I think the approach proposed here is best. I'm biased though as it was my design and I mainly work with pnpm packages and have done this same migration this way before.

**Comment by ChumpChief** [comment]
> Instead of being incremental we could force everyone to upgrade node at the exact same time, and also upgrade npm and the lock file. This seems risky as it would be a large source change (a pain to revert) and likely to break someone (and thus need to be reverted). This seems like a worse option.

I might be missing the risk here - afaik the package-lock format change shouldn't introduce anything beyond changes to the package-lock.json files, and the actual install should remain the same.

Also IMO we should move straight to 18.  16 is already in maintenance and will go out of service in under 6 months (and we'll have to do this again).

**Review by CraigMacomber — APPROVED** [review]

**Review by tylerbutler — APPROVED** [review]
